metapool/plate.py

```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
+        self.data = {'1': None, '2': None, '3': None, '4': None}
+
+    def _get_quadrant(self, quadrant):
+        d = OrderedDict()
+        for i in range(1, 9):
+            row_96 = PlateReplication.row_letters[i - 1]
+            if quadrant in ['1', '2']:
+                row_384 = PlateReplication.row_letters[(2 * i) - 2]
+            else:
+                row_384 = PlateReplication.row_letters[(2 * i) - 1]
+
+            for j in range(1, 13):
+                col_96 = j
+                if quadrant in ['1', '3']:
+                    col_384 = (2 * j) - 1
+                else:
+                    col_384 = (2 * j)
+
+                k = "%s%s" % (row_96, col_96)
+                v = "%s%s" % (row_384, col_384)
+                d[k] = v
+
+        return d
+
+    def get_384_well_location(self, well_96_id, quadrant):
+        '''
+        Translate a 96-well plate + a quadrant into a 384-well plate cell
+        :param well_96_id: A 96-well plate ID
+        :param quadrant: A quadrant of a 384-well plate e.g. '1', '2', '3', '4'
+        :return: A 384-well plate ID
+        '''
+        quadrant = str(quadrant)
+        if quadrant in self.map_to_384:
+            if well_96_id in self.map_to_384[quadrant]:
+                return self.map_to_384[quadrant][well_96_id]
+
+    def get_96_well_location_and_quadrant(self, well_384_id):
+        '''
+        Translate a 384-well plate ID to a 96-well plate ID and a quadrant
+        :param well_384_id: A 384-well plate ID
+        :return: A tuple of (quadrant, 96-well plate ID)
+        '''
+        # not an optimal search but mean-search-space is 192 and won't grow.
+        for quadrant in PlateReplication.quadrants:
+            for k in self.map_to_384[quadrant]:
+                if self.map_to_384[quadrant][k] == well_384_id:
+                    return quadrant, k
+
+    def _map_quadrants(self, src_quad, dst_quad):
+        # since all values in self.d are generated in the same order by
+        # the same function and saved as OrderedDicts, the 384-well locations
+        # from any src quadrant will map directly to those of the dst quad.
+        src = list(self.map_to_384[str(src_quad)].values())
+        dst = list(self.map_to_384[str(dst_quad)].values())
+
+        return dict(map(lambda i, j: (i, j), src, dst))
+
+    def _get_all_384_locations(self, quadrant):
+        return list(self.map_to_384[quadrant].values())
+
+    def _get_quadrants(self, wells_384):
+        results = []
+
+        for i in range(1, 5):
+            all_wells_in_quadrant = set(self._get_all_384_locations(str(i)))
+            if set(wells_384) & all_wells_in_quadrant:
+                results.append(str(i))
+
+        return results
+
+    def check_bounds_384(self, locations):
+        '''
+        Check if one or more 384-well plate IDs are valid.
+        :param locations: A list of one or more locations
+        :return: A list of valid locations or None if none were valid.
```

Comment:
This goes back to my previous comment, is it really safe to return None vs. an empty array?

Reply:
I can go either way. I included the functionality for the wetlab, but I'm not sure if they'll still need to use it since PlateReplicant handles all of the checking for them. I agree it's safer for them to get a list returned.